fix: resolve CORS issues via reverse proxy configuration (Vite, Nginx, Apache)

abhishekblue opened the pull request:
**What kind of change does this PR introduce?**

- Bug fix (non-breaking change which fixes an issue)
- Documentation update

**Issue Number:**

Fixes #4697 

**If relevant, did you update the documentation?**

Yes
[](https://github.com/PalisadoesFoundation/talawa-admin/blob/main/docs/docs/docs/getting-started/configuration.md
)

**Summary**

This PR resolves the CORS and XSS errors encountered when running `talawa-admin` and `talawa-api` on different IP addresses or ports.

As per issue's description, this solution implements a Reverse Proxy architecture directly within `talawa-admin`:

**Development**: Configured vite.config.ts to proxy /graphql requests to the API.

**Production**: Updated nginx.conf and added apache.conf to handle proxying in containerized/server environments.

**Configuration**: Updated .env.example to use relative paths (/graphql) by default.

**Documentation**: Added specific configuration instructions to the Getting Started guide.

**Does this PR introduce a breaking change?**

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

No

## Checklist

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass


**Other information**

Legacy Code Observation: I noticed that `config/docker/setup/nginx.prod.conf` exists but appears to be unused (as Dockerfile.prod explicitly copies nginx.conf). I left it untouched to keep this PR focused, but it is likely dead code that can be removed in a future cleanup.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Chores**
  * Updated webserver configs to proxy GraphQL to the backend, enable WebSocket upgrades, and serve the frontend with an index fallback; removed a previous production webserver config and an explicit CORS block.
  * Simplified dev proxy routing and adjusted proxy targets.

* **Refactor**
  * Converted build tooling to a mode-aware factory for runtime env loading, dynamic ports, and consolidated server/build/plugin configuration.
  * Centralized authorization header handling and changed initial backend probe to a GraphQL POST.

* **Tests**
  * Added unit tests for build-config utilities and expanded test discovery.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

palisadoes commented on config/docker/setup/apache.conf:
```diff
@@ -0,0 +1,21 @@
+<VirtualHost *:80>
+    ServerName localhost
+    DocumentRoot /usr/share/nginx/html
```
1. Why was this added?
2. Why does it refer to nginx?

palisadoes commented on .env.example:
```diff
@@ -8,7 +8,7 @@ PORT=4321
 # Run Talawa-api locally in your system, and put its url into the same.
 
-REACT_APP_TALAWA_URL=http://localhost:4000/graphql
+REACT_APP_TALAWA_URL=/graphql
```
IP addresses must `ALWAYS` be provided in network related parameters.

palisadoes reviewed (changes requested):
See comments

abhishekblue commented on docs/docs/docs/getting-started/configuration.md:
````diff
@@ -68,6 +68,15 @@ If you are trying to access Talawa Admin from a remote host with the API URL con
 ```
+:::tip Proxy Configuration
+To avoid CORS issues when hosting the Admin and API on different IP addresses, you can use a **relative path** for the backend URL.
+
````
I have updated the documentation to reflect the distributed system perspective, clarifying that the Web Server must be configured to proxy traffic to the API IP.

abhishekblue commented:
@palisadoes 
Thanks for the checklist. I have already tested this on my Network IP (192.168.x.x) to make sure it works in a real setup, not just localhost.

Here is how I tested it:
**Setup:** I configured the Admin app `.env` to use the relative path `/graphql`.
**Access:** I accessed the Admin app using my LAN IP (http://192.168.1.10:4321) in the browser.

**Result:**
**Before Fix:** I was getting CORS errors because the IP (4321) did not match the API (4000).
**After Fix:** The requests were successfully proxied to the API, and the app loaded perfectly without any CORS errors.

VanshikaSabharwal commented:
@palisadoes it looks good to me but we have apache server running the demo site, so do we need to have 2 reverse proxies?

palisadoes reviewed (changes requested):
1. This is my configuration. 
    1. My Admin app is running on local host
    2. The API app is running on a completely different machine at 192.168.1.100

        ```
        REACT_APP_BACKEND_WEBSOCKET_URL=ws://192.168.1.100:4000/graphql
        REACT_APP_TALAWA_URL=http://192.168.1.100:4000/graphql
        TALAWA_API_URL=http://192.168.1.100:4000
        ```

1. This is the error I get

    <img width="2131" height="1087" alt="image" src="https://github.com/user-attachments/assets/971658a6-f598-4f82-aa11-4f48ef39f4d7" />

1. I'm running the app without docker

Ideally the solution must:

1. Have only one parameter to be set to access the API. We now have three. The easiest to use would be this as it's already in the setup script and our documentation:
    ```
    REACT_APP_TALAWA_URL
    ```

1. All other API URLs must be computed from that single URL

---

When this works on Admin without Docker, then I'll test on a machine using Docker.

abhishekblue commented:
@palisadoes I have already fixed this. I think the bot comment is just a checklist and I have completed all those points.

For Development: `vite.config.ts` now automatically handles the proxy.

For Production: I updated the `nginx.conf` (used by Docker). So the app works out-of-the-box without needing any external software.

I have verified this on my local network and it is working fine.

abhishekblue commented:
> It's failing in this scenario
> 
> <img alt="image" width="1041" height="1179" src="https://private-user-images.githubusercontent.com/16875803/523892272-6dc8e625-12a2-4efe-a1c4-8d019c1ba94e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjUyNTg4NDgsIm5iZiI6MTc2NTI1ODU0OCwicGF0aCI6Ii8xNjg3NTgwMy81MjM4OTIyNzItNmRjOGU2MjUtMTJhMi00ZWZlLWExYzQtOGQwMTljMWJhOTRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjA5VDA1MzU0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFhODU4NDU1YWU4NzAwZGQ5NDIyZWI0MTBhNDVmZDhhNjUyMzNiZDU4MDNhMjZjNDM3MGI1NDI0NWRiZjE1OGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mtl8j1bXvf0INJ5KePvzmwd7yu81l7NNgCfy6XUytE0">

I think the issue is this: 

REACT_APP_TALAWA_URL was set to `/graphql` (This caused the proxy to default to `localhost`, which failed to reach your API at 192.168.x.x).

You simply set `REACT_APP_TALAWA_URL` to your Full API URL (e.g., http://192.168.1.100:4000/graphql). It will automatically extracts the IP for the proxy.

palisadoes commented:
1. The web socket variable is used in the code to set a partner in index.tsx. it's not obsolete. 
1. You'll need to figure out a way to derive it in your code
1. Also in the docs explain what specific webserver parameter values will be to be adjusted in each of the scenarios. We need to maximize clarity

palisadoes commented:
@abhishekblue 

How can we treat the vite config for as an exception? It's setting up important operational parameters. Is this sometime that is normally done?

abhishekblue commented:
@coderabbitai Please check my [comment](https://github.com/PalisadoesFoundation/talawa-admin/pull/4880#issuecomment-3638598267)  and review my latest commit.

abhishekblue commented:
@coderabbitai

1. <img width="277" height="64" alt="image" src="https://github.com/user-attachments/assets/35087f15-0eed-40a2-a81a-9c2c5cdf1a19" />

2. <img width="516" height="26" alt="image" src="https://github.com/user-attachments/assets/dfc94a07-3e8d-495f-accc-348d7d35dae3" />

3. - Payload tab: `{"operationName":"CurrentUser","variables":{},"query":"query CurrentUser {\n  currentUser {\n    addressLine1\n    addressLine2\n    avatarMimeType\n    avatarURL\n    birthDate\n    city\n    countryCode\n    createdAt\n    description\n    educationGrade\n    emailAddress\n    employmentStatus\n    homePhoneNumber\n    id\n    isEmailAddressVerified\n    maritalStatus\n    mobilePhoneNumber\n    name\n    natalSex\n    naturalLanguageCode\n    postalCode\n    role\n    state\n    updatedAt\n    workPhoneNumber\n    eventsAttended {\n      id\n      __typename\n    }\n    __typename\n  }\n}"}`
- content-type: application/json
- Request URL: http://192.168.1.15:4321/graphql or http://localhost/graphql (based on what I use in browser)

But in the console I am getting this error:
<img width="475" height="47" alt="image" src="https://github.com/user-attachments/assets/e39aa609-1871-4f6b-bb84-8c629e4a0178" />

I have also made a new commit. Please check!

abhishekblue commented on config/vite.config.ts:
```diff
@@ -1,65 +1,132 @@
+    apiTarget = urlObj.origin;
+  } catch {
+    apiTarget = 'http://localhost:4000';
+  }
```
There is no `TALAWA_API_URL`. Both docs and `.env.example` mentions `REACT_APP_TALAWA_URL`

palisadoes commented:
Is this ready to test?

abhishekblue commented:
> Is this ready to test?

Not yet, trying to resolve this "body dropping" issue. 
I think rest is done, this is the only thing that's left.

abhishekblue commented:
@Coderabbitai, please have a look at `configuration.md` as well.

palisadoes reviewed (changes requested):
@abhishekblue 

1. You need to fix the conflicting file. It cannot be merged without fixing it.
2. Is this ready to be reviewed?

palisadoes commented:
Please fix the failing test
